<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="combiroc">
<title>Signature refining tutorial - scRNA-seq Workflow • combiroc</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Signature refining tutorial - scRNA-seq Workflow">
<meta property="og:description" content="combiroc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">combiroc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/combiroc_vignette_1.html">Standard Workflow</a>
    <a class="dropdown-item" href="../articles/combiroc_vignette_2.html">SingleCell Workflow</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ingmbioinfo/combiroc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Signature refining tutorial - scRNA-seq Workflow</h1>
            <h3 data-toc-skip class="subtitle"><div class="line-block">Combiroc-driven optimization of NK cells marker
signature at single cell resolution.</div></h3>
                        <h4 data-toc-skip class="author">Riccardo L.
Rossi, Ivan Ferrari</h4>
            
            <h4 data-toc-skip class="date">30 June 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ingmbioinfo/combiroc/blob/HEAD/vignettes/combiroc_vignette_2.Rmd" class="external-link"><code>vignettes/combiroc_vignette_2.Rmd</code></a></small>
      <div class="d-none name"><code>combiroc_vignette_2.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="top">Introduction<a class="anchor" aria-label="anchor" href="#top"></a>
</h2>
<p>Single-cell transcriptomics is a rapidly growing field which is
essential to improve our understanding of complex tissues and organisms
at an unprecedented resolution. The manual annotation of cell identity,
which is the gold standard procedure to determine cell populations,
remains a major limitation in most scRNA-seq analysis <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1795-z" class="external-link">[1]</a>;
more recently, a growing number of resources are being published
offering the scientific community many cell type specific gene signature
as a reference. Alas, marker gene belonging to gene signatures could be
unspecific if taken singularly, specially if that given gene is involved
in all the differentiation steps in some cellular population of the
dataset (from stem cells to a terminally differentiated population),
making it a differentially expressed gene (DEG) in more than one
cluster. In this context we decided to enlarge combiroc’s breath to the
scRNA-seq analysis field as a <strong>signature refiner</strong>, in
order to find the best <strong>few</strong> marker genes combinations
from an <em>already existing</em> gene signature, <em>regardless the
genes’ ranking</em> as differentially expressed. Combiroc is specialized
in finding specific combinations, thus its use may allow researchers to
find some cells of interest with a high degree of confidence and help in
the process of manual annotation by reducing the number of marker genes
to consider.</p>
<p>Here we show the complete workflow of combiroc-mediated signature
refinement at single-cell resolution in which we used 3 PBMC public
datasets showed in <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>
vignettes. All of them have been deeply characterized and scrupulously
labelled, so they are accepted and used as reference by the scientific
community.</p>
</div>
<div class="section level2">
<h2 id="libraries-needed">Libraries needed<a class="anchor" aria-label="anchor" href="#libraries-needed"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># main libraries:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ingmbioinfo/combiroc" class="external-link">combiroc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>and also needed for this specific scRNAseq workflow (install them
first, if you didn’t already):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span> <span class="co"># used to install devel-level packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># used to process scRNAseq data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratData</span><span class="op">)</span> <span class="co"># used to install and load scRNAseq datasets</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratDisk</span><span class="op">)</span> <span class="co"># used to read h5seurat files</span></span></code></pre></div>
<p><strong>A note on executing this vignette</strong></p>
<p>The code described in this vignette was originally run on a
jupyter-based R-kernel notebook backed by a high-performance computing
(HPC) server infrastructure. Due to relevant size of some of the objects
loaded and/or created, <strong>it is unlikely that the whole code can be
run on a standard computing equipment (i.e. a laptop)</strong>. The
computationally heavy passages are therefore marked as <span style="color: orange;">[Heavy-Code]</span> and can be skipped by the
user by reading/loading <strong>pre-computed objects or intermediate
results either from <code>rds</code> or <code>csv</code> files</strong>.
In this way the reader/reviewer will be able to complete the whole
workflow without having a HPC server.</p>
</div>
<div class="section level2">
<h2 id="selection-of-nk-cells-marker-combinations">Selection of NK cells marker combinations<a class="anchor" aria-label="anchor" href="#selection-of-nk-cells-marker-combinations"></a>
</h2>
<p>The purpose of this analysis is to find the best combinations of
known gene markers that best describe the NK-cell populations starting
from a NK-cell single-cell transcriptomic signature, then using the
models calculated on such combinations on other scRNAseq datasets.<br>
The difficulty to distinguish NK (Natural Killer) cells from CD8+ T
cells in many scRNAseq datasets is a good challenge to test combiroc
performance in markers signature optimization. After performing
differential marker expression on an initial “training” scRNAseq dataset
we will take into account the top 30 differentially expressed genes
(DEGs) specific for NK cells cluster; then we will use combiroc to
select the best gene combinations among these top 30 NK cells markers.
Regression models from the selected combinations will then be used on
other independent datasets (“test” datasets) of the same kind.</p>
<p>As combiroc training dataset we are using the <a href="https://satijalab.org/seurat/articles/multimodal_reference_mapping.html" class="external-link">PBMC
CITE-seq atlas</a> from Hao et al. 2021 <a href="https://www.sciencedirect.com/science/article/pii/S0092867421005833" class="external-link">[2]</a>.
This dataset can be downloaded as an h5 Seurat data (h5s) from the <a href="https://atlas.fredhutch.org/nygc/multimodal-pbmc/" class="external-link">FredHutch-NYCG
atlas page</a> (exact download link: <a href="https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat" class="external-link">here</a>)
<span style="color: orange;">[Heavy-Code]</span></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read the downloaded h5seurat dataset file (using SeuratDisk functions)</span></span>
<span><span class="va">atlas</span> <span class="op">&lt;-</span> <span class="fu">LoadH5Seurat</span><span class="op">(</span><span class="st">"pbmc_multimodal.h5seurat"</span><span class="op">)</span></span>
<span><span class="co"># activate the level-1 annotations</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">atlas</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">atlas</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">celltype.l1</span></span>
<span><span class="co"># overview annotated cell clusters with UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">atlas</span>, label <span class="op">=</span> <span class="cn">T</span>, repel <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="atlas_dimplot.png" width="31104" style="display: block; margin: auto;"></p>
<p>To guarantee the best possible performance of the training dataset
(the one we extract the markers combinations from) is important to make
sure its value ranges are similar to those we wanto to use as
test/validation datasets. For this reason we advise using the raw counts
matrices for all the datasets, in order to be able to consistently
rescale them in the same way by using the function
<code><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">Seurat::ScaleData()</a></code>.</p>
<div class="section level3">
<h3 id="extracting-the-nk-markers">Extracting the NK markers<a class="anchor" aria-label="anchor" href="#extracting-the-nk-markers"></a>
</h3>
<p>The gene markers for NK cells were obtained with a standard Seurat
analysis (see <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">here</a>);
the <code><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">Seurat::FindMarkers()</a></code> function was used, then markers
were ordered by fold change: <span style="color: orange;">[Heavy-Code]</span></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Performing differential expression analysis</span></span>
<span><span class="va">nk.de.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">atlas</span>, ident.1 <span class="op">=</span> <span class="st">"NK"</span>, ident.2 <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">nk.de.markers</span> <span class="op">&lt;-</span> <span class="va">nk.de.markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">nk.de.markers</span><span class="op">$</span><span class="va">avg_log2FC</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">nk_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">nk.de.markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span></span></code></pre></div>
<p>By visualising the expression values of the top 4 markers of NK cells
cluster it is evident that the majority of them is also higly expressed
in other non-NK clusters (e.g. CD8 T), making them too unspecific to
allow a well defined cluster annotation if considered individually.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">atlas</span>, <span class="va">nk_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="FeaturePlot1.png" width="36000" style="display: block; margin: auto;"></p>
<p>The user not willing (or computationally limited) to run the
differential expression analysis above can read the pre-computed object
containing the gene signature of the <strong>top 30 DEGs</strong>
(overexpressed in NK cells vs all others):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nk.de.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">'inst/precomp/nk_degs'</span><span class="op">)</span></span>
<span><span class="va">nk_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">nk.de.markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span></span>
<span><span class="va">nk_genes</span></span>
<span><span class="co">#&gt;  [1] "GNLY"    "NKG7"    "GZMB"    "PRF1"    "FGFBP2"  "GZMA"    "KLRD1"  </span></span>
<span><span class="co">#&gt;  [8] "CST7"    "SPON2"   "KLRF1"   "CTSW"    "CD247"   "CCL5"    "CLIC3"  </span></span>
<span><span class="co">#&gt; [15] "HOPX"    "GZMH"    "IL2RB"   "KLRB1"   "TRDC"    "CD7"     "GZMM"   </span></span>
<span><span class="co">#&gt; [22] "MYOM2"   "FCGR3A"  "ARL4C"   "ABHD17A" "SYNE2"   "CMC1"    "EFHD2"  </span></span>
<span><span class="co">#&gt; [29] "ADGRG1"  "JAK1"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="converting-seurat-objects-in-combirocs-input-format">Converting Seurat objects in combiroc’s input format<a class="anchor" aria-label="anchor" href="#converting-seurat-objects-in-combirocs-input-format"></a>
</h3>
<p>Central to the combiroc workflow applied to single cell data is the
function <code><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc()</a></code>. It takes a Seurat object as
input and extracts both the selected marker expression values from the
<span class="citation">@data</span> slot of a given assay and,
optionally, the class of any specific celltype (in our case ‘NK’ or
‘Other’); then it directly assembles a dataframe compatible with
combiroc workflow either for training (finding combinations and models)
or testing purposes (using previously found combinations).</p>
</div>
<div class="section level3">
<h3 id="preparing-the-training-single-cell-rnaseq-dataset">Preparing the training single-cell RNAseq dataset<a class="anchor" aria-label="anchor" href="#preparing-the-training-single-cell-rnaseq-dataset"></a>
</h3>
<p>Once we have a Seurat object obtained from a standard Seurat
protocol, it needs to be converted in a combiroc-ready version in order
to perform the training procedure, i.e. finding the best markers
combinations and models, with combiroc’s functions. Such a
<code>train</code> object can be obtained with the
<code><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc()</a></code> function: <span style="color: orange;">[Heavy-Code]</span></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc</a></span><span class="op">(</span>SeuratObject <span class="op">=</span> <span class="va">atlas</span>, </span>
<span>                            gene_list <span class="op">=</span> <span class="va">nk_genes</span>, </span>
<span>                            assay <span class="op">=</span> <span class="st">'SCT'</span>, </span>
<span>                            labelled_data <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                            case_class <span class="op">=</span> <span class="st">'NK'</span>, </span>
<span>                            case_label <span class="op">=</span> <span class="st">'NK'</span>, </span>
<span>                            control_label <span class="op">=</span>  <span class="st">'Other'</span><span class="op">)</span></span></code></pre></div>
<p>The pre-computed <code>train</code> object obtained with the code
above can be directly read here:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"inst/precomp/train.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">train</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    ID Class   ABHD17A    ADGRG1    ARL4C      CCL5     CD247</span></span>
<span><span class="co">#&gt; 1 L1_AAACCCAAGAAACTCA Other 0.6931472 0.0000000 0.000000 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; 2 L1_AAACCCAAGACATACA Other 0.0000000 0.0000000 1.609438 0.0000000 0.6931472</span></span>
<span><span class="co">#&gt; 3 L1_AAACCCACAACTGGTT Other 0.0000000 0.0000000 0.000000 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; 4 L1_AAACCCACACGTACTA    NK 1.3862944 0.6931472 1.609438 2.4849066 1.7917595</span></span>
<span><span class="co">#&gt; 5 L1_AAACCCACAGCATACT Other 0.0000000 0.0000000 1.386294 0.6931472 0.6931472</span></span>
<span><span class="co">#&gt; 6 L1_AAACCCACATCAGTCA Other 0.0000000 0.0000000 2.302585 2.9957323 1.6094379</span></span>
<span><span class="co">#&gt;         CD7    CLIC3     CMC1     CST7     CTSW     EFHD2   FCGR3A   FGFBP2</span></span>
<span><span class="co">#&gt; 1 0.0000000 0.000000 0.000000 0.000000 0.000000 1.0986123 0.000000 0.000000</span></span>
<span><span class="co">#&gt; 2 1.6094379 0.000000 0.000000 0.000000 0.000000 0.6931472 0.000000 0.000000</span></span>
<span><span class="co">#&gt; 3 0.6931472 0.000000 0.000000 0.000000 0.000000 0.0000000 0.000000 0.000000</span></span>
<span><span class="co">#&gt; 4 1.6094379 2.197225 1.945910 2.944439 1.945910 1.3862944 1.386294 2.079442</span></span>
<span><span class="co">#&gt; 5 0.6931472 0.000000 0.000000 0.000000 1.098612 0.6931472 0.000000 0.000000</span></span>
<span><span class="co">#&gt; 6 0.0000000 0.000000 2.302585 2.079442 0.000000 0.0000000 0.000000 0.000000</span></span>
<span><span class="co">#&gt;       GNLY     GZMA     GZMB    GZMH      GZMM     HOPX    IL2RB      JAK1</span></span>
<span><span class="co">#&gt; 1 0.000000 0.000000 0.000000 0.00000 0.0000000 0.000000 0.000000 0.6931472</span></span>
<span><span class="co">#&gt; 2 0.000000 0.000000 0.000000 0.00000 0.0000000 0.000000 0.000000 1.3862944</span></span>
<span><span class="co">#&gt; 3 0.000000 0.000000 0.000000 0.00000 0.6931472 0.000000 0.000000 0.0000000</span></span>
<span><span class="co">#&gt; 4 3.044522 1.386294 2.772589 1.94591 1.3862944 1.609438 1.791759 1.9459101</span></span>
<span><span class="co">#&gt; 5 0.000000 0.000000 0.000000 0.00000 1.0986123 0.000000 0.000000 0.6931472</span></span>
<span><span class="co">#&gt; 6 0.000000 1.098612 0.000000 0.00000 1.9459101 0.000000 0.000000 1.3862944</span></span>
<span><span class="co">#&gt;      KLRB1    KLRD1    KLRF1    MYOM2      NKG7     PRF1    SPON2     SYNE2</span></span>
<span><span class="co">#&gt; 1 0.000000 0.000000 0.000000 0.000000 0.0000000 0.000000 0.000000 0.0000000</span></span>
<span><span class="co">#&gt; 2 2.079442 0.000000 0.000000 0.000000 0.0000000 0.000000 0.000000 0.0000000</span></span>
<span><span class="co">#&gt; 3 0.000000 0.000000 0.000000 0.000000 0.6931472 0.000000 0.000000 0.6931472</span></span>
<span><span class="co">#&gt; 4 1.386294 1.609438 1.386294 1.791759 3.9120230 1.791759 2.397895 1.0986123</span></span>
<span><span class="co">#&gt; 5 0.000000 0.000000 0.000000 0.000000 0.0000000 0.000000 0.000000 0.0000000</span></span>
<span><span class="co">#&gt; 6 0.000000 0.000000 0.000000 0.000000 2.9957323 0.000000 0.000000 1.7917595</span></span>
<span><span class="co">#&gt;       TRDC</span></span>
<span><span class="co">#&gt; 1 0.000000</span></span>
<span><span class="co">#&gt; 2 0.000000</span></span>
<span><span class="co">#&gt; 3 0.000000</span></span>
<span><span class="co">#&gt; 4 1.791759</span></span>
<span><span class="co">#&gt; 5 0.000000</span></span>
<span><span class="co">#&gt; 6 0.000000</span></span></code></pre></div>
<p>Then, as described in the <a href="https://ingmbioinfo.github.io/combiroc/articles/combiroc_vignette_1.html">main
combiroc vignette</a>, we used <code>combiroc_long</code> to make the
data in long tidy format, fit for further processing:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combiroc_long.html">combiroc_long</a></span><span class="op">(</span><span class="va">train</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">train_long</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   ID                  Class Markers Values</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> L1_AAACCCAAGAAACTCA Other ABHD17A  0.693</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> L1_AAACCCAAGAAACTCA Other ADGRG1   0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> L1_AAACCCAAGAAACTCA Other ARL4C    0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> L1_AAACCCAAGAAACTCA Other CCL5     0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> L1_AAACCCAAGAAACTCA Other CD247    0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> L1_AAACCCAAGAAACTCA Other CD7      0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="finding-the-best-marker-combinations-and-models">Finding the best marker combinations and models<a class="anchor" aria-label="anchor" href="#finding-the-best-marker-combinations-and-models"></a>
</h3>
<p>Given a list of markers, combiroc assesses the performance of all
combinations of such markers. The computational load of this process can
be high for more than 10 markers. A list of <span class="math inline">\(n\)</span> markers generates <span class="math inline">\({2^{n} - 1 }\)</span>, thus for <span class="math inline">\(n=30\)</span> we have <strong><em>more than a
billion</em></strong> possible combinations.<br>
Mathematically, for combinations here we mean <em>combinations without
repetition</em>, i.e. a subset of <span class="math inline">\(k\)</span>
distinct elements of a set with <span class="math inline">\(n\)</span>
elements, that can be calculated as the <a href="https://en.wikipedia.org/wiki/Binomial_coefficient#Binomial_coefficients_as_polynomials" class="external-link">binomial
coefficient</a> <span class="math inline">\(\binom{n}{k}\)</span>.</p>
<p>The purpose of combiroc is always finding the best <em>optimized</em>
combination of markers, i.e. a <em>subset</em> of the original full
signature which, despite the <em>much smaller</em> number of markers,
retains the discriminatory power of the original signature or it’s even
better. This is particularly important in the field of diagnostics where
a smaller set of marker has a bigger translational potential (see our
first combiroc paper <a href="https://www.nature.com/articles/srep45477" class="external-link">Mazzara et al. 2017</a>
for a discussion on this).</p>
<p>Similarly, here we will limit the search to combinations composed by
<strong>no more than 5 markers</strong>: to do so we will set the
<code>max_length = 5</code> attribute of the <code><a href="../reference/combi.html">combi()</a></code>
function. With this limitation the number of combinations to compute
drops to 174,436, which is computationally more manageable.</p>
<p><span class="math display">\[ Combinations(upTo5Markers) =
\sum_{i=1}^5 \binom{n}{i} \]</span></p>
<pre><code><span><span class="co">#&gt;   n_markers n_combs tot_combs</span></span>
<span><span class="co">#&gt; 1         1      30        30</span></span>
<span><span class="co">#&gt; 2         2     435       465</span></span>
<span><span class="co">#&gt; 3         3    4060      4525</span></span>
<span><span class="co">#&gt; 4         4   27405     31930</span></span>
<span><span class="co">#&gt; 5         5  142506    174436</span></span></code></pre>
<p><img src="combiroc_vignette_2_files/figure-html/unnamed-chunk-14-1.png" width="384" style="display: block; margin: auto;"></p>
<p>The <code><a href="../reference/combi.html">combi()</a></code> function works on the <code>train</code>
dataset computing the marker combinations and counting their
corresponding positive samples in each class (once thresholds are
selected).<br>
A sample, to be considered positive for a given combination, must have a
value higher than a given signal threshold (<strong>signalthr</strong>)
for at least a given number of markers composing that combination
(<strong>combithr</strong>).</p>
<p>As described in the <a href="https://ingmbioinfo.github.io/combiroc/articles/combiroc_vignette_1.html">combiroc’s
vignette for the standard workflow</a>, the argument
<strong>signalthr</strong> in the <code><a href="../reference/combi.html">combi()</a></code> function should
be set according to the guidelines and characteristics of the
methodology used for the analysis or by an accurate inspection of the
signal intensity distribution. In the event specific guidelines are
missing, one should set the value signalthr as suggested by the
<code>distr$Density_plot</code> feature.</p>
<p>Single cell RNA-seq datasets are exactly one of such cases: it is
often neither possible nor easy to extrapolate the best signal
threshold, due to the highly scattered distribution of the expression
values: here we thus take advantage of the package-specific
<code>distr$Density_plot</code> feature and will let the package find
this value without the user intervention.</p>
</div>
<div class="section level3">
<h3 id="combinatorial-analysis">Combinatorial analysis<a class="anchor" aria-label="anchor" href="#combinatorial-analysis"></a>
</h3>
<p>The core of a combiroc analysis are the two
<code>marker_distribution()</code> and <code><a href="../reference/combi.html">combi()</a></code> functions:
changing the value of key parameters in these functions we can modulate
the stringency and severity of the analysis.</p>
<p>To obtain a performing and stringent signature refinement we set
<code>combithr</code> argument to <strong>2</strong> in
<code><a href="../reference/combi.html">combi()</a></code>: this is the minimum number of positively expressed
genes (i.e. the minimum number of genes that need to reach the signal
threshold) to consider the whole combination as a valid one.</p>
<p>The plots below show expression values (signal) distribution obtained
with the <code>markers_distributions()</code> functions on the training
dataset in long format (<code>train_long</code>). Differently from two
well-defined classes (as seen in the standard combiroc workflow), it’s
not intuitive where to put the signal threshold, but the function
suggests a signal intensity value of <strong>0.9</strong>. Genes
expression values, as expected, are distributed in different ways in the
two classes:</p>
<p><em>Please note: the <code><a href="../reference/markers_distribution.html">markers_distribution()</a></code> command
naturally triggers a few warnings in which the user is reminded how to
use the command arguments</em></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">distr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markers_distribution.html">markers_distribution</a></span><span class="op">(</span><span class="va">train_long</span>,</span>
<span>                              signalthr_prediction <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                              case_class <span class="op">=</span> <span class="st">"NK"</span><span class="op">)</span></span>
<span><span class="va">distr</span><span class="op">$</span><span class="va">Density_plot</span></span></code></pre></div>
<p><img src="combiroc_vignette_2_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<p>The suggested threshold is indicated by the vertical dashed line and
its value is shown in overlay. The exact value can be retrieved from the
<code>distr</code> object by <code>distr$Coord$threshold</code> (see the
<a href="https://ingmbioinfo.github.io/combiroc/articles/combiroc_vignette_1.html#dens">Standard
workflow</a> for further details):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">autoThreshold</span> <span class="op">&lt;-</span> <span class="va">distr</span><span class="op">$</span><span class="va">Coord</span><span class="op">[</span><span class="va">distr</span><span class="op">$</span><span class="va">Coord</span><span class="op">$</span><span class="va">Youden</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">distr</span><span class="op">$</span><span class="va">Coord</span><span class="op">$</span><span class="va">Youden</span><span class="op">)</span>, <span class="st">'threshold'</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">distr</span><span class="op">$</span><span class="va">Coord</span><span class="op">[</span><span class="va">distr</span><span class="op">$</span><span class="va">Coord</span><span class="op">$</span><span class="va">Youden</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">distr</span><span class="op">$</span><span class="va">Coord</span><span class="op">$</span><span class="va">Youden</span><span class="op">)</span>,<span class="op">]</span> <span class="co"># optimal threshold</span></span>
<span><span class="co">#&gt;   threshold specificity sensitivity    Youden</span></span>
<span><span class="co">#&gt; 3 0.8958797          86          79 0.6461265</span></span></code></pre></div>
<p>Let’s apply now the <code><a href="../reference/combi.html">combi()</a></code> function on the training
dataset <code>train</code> with the suggested threshold. The results is
a table (the <code>train_tab</code> dataframe) of all the combinations
with 1 (= the individual markers) to 5 genes in them.</p>
<p>In this dataframe are also shown sensitivity (<strong>SE</strong>)
and specificity (<strong>SP</strong>) for each combination:</p>
<ul>
<li>
<strong>SE</strong> as the percentage of positive cells for a ‘NK’
class</li>
<li>
<strong>SP</strong> as the percentage of negative cells for ‘Other’
class</li>
</ul>
<p><span style="color: orange;">[Heavy-Code]</span> <em>(use the
pre-computed below)</em></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combi.html">combi</a></span><span class="op">(</span><span class="va">train</span>, signalthr <span class="op">=</span> <span class="va">autoThreshold</span>, combithr <span class="op">=</span> <span class="fl">2</span>, max_length <span class="op">=</span> <span class="fl">5</span>, case_class <span class="op">=</span> <span class="st">'NK'</span><span class="op">)</span></span></code></pre></div>
<p>Read the precomputed object:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">'inst/precomp/train_tab'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">train_tab</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                        Markers X.Positives.Other X.Positives.NK</span></span>
<span><span class="co">#&gt; Combination 174401 MYOM2-NKG7-PRF1-SPON2-SYNE2             20313          18220</span></span>
<span><span class="co">#&gt; Combination 174402  MYOM2-NKG7-PRF1-SPON2-TRDC             12876          18413</span></span>
<span><span class="co">#&gt; Combination 174403  MYOM2-NKG7-PRF1-SYNE2-TRDC             20844          18482</span></span>
<span><span class="co">#&gt; Combination 174404 MYOM2-NKG7-SPON2-SYNE2-TRDC             19664          18206</span></span>
<span><span class="co">#&gt; Combination 174405 MYOM2-PRF1-SPON2-SYNE2-TRDC             12019          17767</span></span>
<span><span class="co">#&gt; Combination 174406  NKG7-PRF1-SPON2-SYNE2-TRDC             20866          18512</span></span>
<span><span class="co">#&gt;                          SE       SP n_markers</span></span>
<span><span class="co">#&gt; Combination 174401 97.62109 85.80503         5</span></span>
<span><span class="co">#&gt; Combination 174402 98.65517 91.00210         5</span></span>
<span><span class="co">#&gt; Combination 174403 99.02486 85.43396         5</span></span>
<span><span class="co">#&gt; Combination 174404 97.54608 86.25856         5</span></span>
<span><span class="co">#&gt; Combination 174405 95.19396 91.60098         5</span></span>
<span><span class="co">#&gt; Combination 174406 99.18560 85.41859         5</span></span></code></pre></div>
<p>Then we rank the combinations by Youden index (SE+SP-1) and we filter
them by their SE and SP values. We will select the top 4 (for
demonstration purposes only, others may be also good): what we are using
to rank is the Youden index and the number of genes in the combination,
where combinations with fewer genes are considered better ones.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rmks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ranked_combs.html">ranked_combs</a></span><span class="op">(</span>combo_table <span class="op">=</span> <span class="va">train_tab</span>, </span>
<span>                     min_SE <span class="op">=</span> <span class="fl">95</span>, </span>
<span>                     min_SP <span class="op">=</span> <span class="fl">95</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rmks</span><span class="op">$</span><span class="va">table</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                           Markers X.Positives.Other</span></span>
<span><span class="co">#&gt; Combination 169807    GNLY-IL2RB-KLRF1-SPON2-TRDC              6914</span></span>
<span><span class="co">#&gt; Combination 172173    GZMB-IL2RB-KLRF1-SPON2-TRDC              5633</span></span>
<span><span class="co">#&gt; Combination 163878  FCGR3A-GNLY-IL2RB-KLRF1-MYOM2              6401</span></span>
<span><span class="co">#&gt; Combination 137550 CLIC3-FCGR3A-IL2RB-KLRD1-KLRF1              6777</span></span>
<span><span class="co">#&gt; Combination 138800   CLIC3-GNLY-IL2RB-KLRF1-SPON2              6361</span></span>
<span><span class="co">#&gt; Combination 164702   FCGR3A-GZMB-IL2RB-KLRF1-TRDC              6351</span></span>
<span><span class="co">#&gt;                    X.Positives.NK       SE       SP n_markers    Youden</span></span>
<span><span class="co">#&gt; Combination 169807          18356 98.34976 95.16841         5 0.9351818</span></span>
<span><span class="co">#&gt; Combination 172173          18186 97.43892 96.06359         5 0.9350251</span></span>
<span><span class="co">#&gt; Combination 163878          18276 97.92113 95.52690         5 0.9344804</span></span>
<span><span class="co">#&gt; Combination 137550          18325 98.18367 95.26415         5 0.9344782</span></span>
<span><span class="co">#&gt; Combination 138800          18268 97.87827 95.55486         5 0.9343313</span></span>
<span><span class="co">#&gt; Combination 164702          18264 97.85684 95.56184         5 0.9341868</span></span></code></pre></div>
<p>And with <code><a href="../reference/roc_reports.html">roc_reports()</a></code> we carry out the computations for
the top four combinations (we need to explicitly indicate their number
IDs in the <code>selected_combinations</code> argument). Also, we need
to indicate the case class (NK):</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reports</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/roc_reports.html">roc_reports</a></span><span class="op">(</span><span class="va">train</span>, </span>
<span>                      markers_table <span class="op">=</span> <span class="va">train_tab</span>, </span>
<span>                      selected_combinations <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">169807</span>,<span class="fl">172173</span>,<span class="fl">163878</span>, <span class="fl">137550</span><span class="op">)</span>,                      </span>
<span>                      case_class <span class="op">=</span> <span class="st">'NK'</span>,</span>
<span>                      single_markers<span class="op">=</span> <span class="va">nk_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Let’s explore the results generated:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reports</span><span class="op">$</span><span class="va">Plot</span></span></code></pre></div>
<p><img src="combiroc_vignette_2_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Zooming in to show the higher performance of the combinations:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rocplot</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">reports</span><span class="op">$</span><span class="va">Plot</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rocplot</span></span></code></pre></div>
<p><img src="combiroc_vignette_2_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reports</span><span class="op">$</span><span class="va">Metrics</span></span>
<span><span class="co">#&gt;                      AUC    SE    SP CutOff   ACC     TN    TP   FN    FP   NPV</span></span>
<span><span class="co">#&gt; GNLY               0.962 0.959 0.914  0.146 0.920 130861 17908  756 12239 0.994</span></span>
<span><span class="co">#&gt; NKG7               0.970 0.982 0.901  0.131 0.911 128972 18336  328 14128 0.997</span></span>
<span><span class="co">#&gt; GZMB               0.949 0.928 0.925  0.136 0.925 132375 17316 1348 10725 0.990</span></span>
<span><span class="co">#&gt; PRF1               0.965 0.935 0.919  0.137 0.921 131569 17452 1212 11531 0.991</span></span>
<span><span class="co">#&gt; Combination 169807 0.995 0.982 0.968  0.107 0.970 138553 18326  338  4547 0.998</span></span>
<span><span class="co">#&gt; Combination 172173 0.995 0.977 0.970  0.108 0.971 138769 18229  435  4331 0.997</span></span>
<span><span class="co">#&gt; Combination 163878 0.995 0.985 0.962  0.078 0.965 137680 18393  271  5420 0.998</span></span>
<span><span class="co">#&gt; Combination 137550 0.995 0.976 0.966  0.101 0.967 138179 18224  440  4921 0.997</span></span>
<span><span class="co">#&gt;                      PPV</span></span>
<span><span class="co">#&gt; GNLY               0.594</span></span>
<span><span class="co">#&gt; NKG7               0.565</span></span>
<span><span class="co">#&gt; GZMB               0.618</span></span>
<span><span class="co">#&gt; PRF1               0.602</span></span>
<span><span class="co">#&gt; Combination 169807 0.801</span></span>
<span><span class="co">#&gt; Combination 172173 0.808</span></span>
<span><span class="co">#&gt; Combination 163878 0.772</span></span>
<span><span class="co">#&gt; Combination 137550 0.787</span></span></code></pre></div>
<p>The overall internal classification performances of the top 4
combinations are better than the top 4 single markers ones, and are very
similar with each other.</p>
<p>To know which are the genes in any specific combination we can use
the <code><a href="../reference/show_markers.html">show_markers()</a></code> function. For combinations 169807 and
172173, use:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/show_markers.html">show_markers</a></span><span class="op">(</span>selected_combinations <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">169807</span>,<span class="fl">172713</span><span class="op">)</span>, markers_table <span class="op">=</span> <span class="va">train_tab</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Combination           Composing_markers</span></span>
<span><span class="co">#&gt; 1 Combination 169807 GNLY-IL2RB-KLRF1-SPON2-TRDC</span></span>
<span><span class="co">#&gt; 2 Combination 172713 GZMH-HOPX-KLRB1-KLRF1-MYOM2</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="testing-the-combinations-on-independent-datasets-validation">Testing the combinations on independent datasets (validation)<a class="anchor" aria-label="anchor" href="#testing-the-combinations-on-independent-datasets-validation"></a>
</h2>
<p><strong>TEST/VALIDATION DATASETS</strong></p>
<p>As independent test/validation we used 2 different datasets on which
we will evaluate the expression levels of the previously selected NK
related marker combinations in each cluster of the datasets:</p>
<ul>
<li>Cord Blood mononuclear cells (CBMC) CITE-seq dataset shown in <a href="https://satijalab.org/seurat/articles/multimodal_vignette.html" class="external-link">‘Using
Seurat with multimodal data’</a> vignette <a href="https://www.nature.com/articles/nmeth.4380" class="external-link">[3]</a>
</li>
<li>Peripheral blood mononuclear cells (PBMC) <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">PBMC3K
scRNA-seq</a> dataset from Satija, R., Farrell, J., Gennert, D. et
al. 2015 <a href="https://www.nature.com/articles/nbt.3192" class="external-link">[4]</a>
</li>
</ul>
<p><strong>Please note</strong>: the main reason we choose the above
datasets is because <strong>they are well annotated</strong>. While the
cell clusters annotation of the first (training) dataset was used in the
computations by combiroc to distinguish NK cells from all other cells,
the cells annotations of the other two validation datasets are taken
away from the data, and <strong>they will be only used <em>ex-post</em>
as a ground truth to check the expression of marker
combinations</strong>.</p>
<div class="section level3">
<h3 id="testing-marker-combinations-on-cbmc-dataset">Testing marker combinations on CBMC dataset<a class="anchor" aria-label="anchor" href="#testing-marker-combinations-on-cbmc-dataset"></a>
</h3>
<p>The <strong>CBMC test dataset</strong>, made of 8617 cord blood
mononuclear cells (CBMCs), produced with CITE-seq was directly loaded
from the <code>SeuratData</code> package. The dataset has been generated
by Stoeckius et al. 2017 <a href="https://www.nature.com/articles/nmeth.4380" class="external-link">[3]</a>. Further
description can be found in the <a href="https://satijalab.org/seurat/articles/multimodal_vignette.html" class="external-link">Seurat
multimodal vignette</a> or by calling the guide with
<code>?cbmc</code>.</p>
<p>We are going to invoke the cbmc dataset from the
<code>SeuratData</code> package and preprocess it as needed:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve cbmc data from SeuratData package</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"SeuratData"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'satijalab/seurat-data'</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratData</span><span class="op">)</span></span>
<span><span class="fu">InstallData</span><span class="op">(</span><span class="st">"cbmc.SeuratData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># process data with standard Seurat protocol</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">cbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">cbmc</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">cbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">cbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">cbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">cbmc</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">cbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">cbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add cell annotations from CITE seq to identities</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">cbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">protein_annotations</span></span>
<span><span class="va">cbmc</span><span class="op">[[</span><span class="st">'ID'</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">cbmc</span><span class="op">)</span></span>
<span><span class="va">cbmc</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize UMAP clusters</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, repel <span class="op">=</span> <span class="cn">T</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="cbmc_dimplot.png" width="67200" style="display: block; margin: auto;"></p>
<p>Please note that the “NA” gray cells cluster is from mouse cells
whose presence acts as negative control in the original
experiment/dataset. Also, cluster annotations are just superimposed and
pulled out as ground truth of cells identity but are not further used in
the combiroc procedure.</p>
<p>As previosuly done for the training dataset, we now obtain the first
combiroc test dataset using <code><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc()</a></code> function.
The <code>test_cbmc</code> object contains cell barcodes as sample IDs
and the expression values of the 30 NK signature genes).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_cbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc</a></span><span class="op">(</span>SeuratObject <span class="op">=</span> <span class="va">cbmc</span>, </span>
<span>                                gene_list <span class="op">=</span> <span class="va">nk_genes</span>, </span>
<span>                                assay <span class="op">=</span> <span class="st">'RNA'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test_cbmc</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="distribution-of-individual-markers-expression">Distribution of individual markers’ expression<a class="anchor" aria-label="anchor" href="#distribution-of-individual-markers-expression"></a>
</h4>
<p>Let’s have a look at how the top four single NK-markers expression is
distributed across the cell clusters in the CBMC dataset:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features <span class="op">=</span> <span class="va">nk_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="cbmc_violinplots_genes.png" width="67200" style="display: block; margin: auto;"></p>
<p>In this dataset, expressions of top individual NK markers is higher
in the NK cluster, but also moderate in others: they are expressed in
some CD8 T cells, CD34+ and CD16+ Mono cells, even if clusters have been
defined and annotated relying on protein-level information (CITE seq).
Moreover, GZMB alone is not sufficient to discriminate NK cells from
pDCs. Only PRF1 among the top four markers is able to discriminate NK
cells alone. Mouse cells (the “NA” cluster), which are kept from the
original dataset as negative control, also show a bit of GNLY and NKG7
expression, which is not surprising given the mixed nature of this
cluster.</p>
</div>
<div class="section level4">
<h4 id="combiroc-nk-probability-across-cbmc-cell-clusters">Combiroc NK probability across CBMC cell clusters<a class="anchor" aria-label="anchor" href="#combiroc-nk-probability-across-cbmc-cell-clusters"></a>
</h4>
<p>We now look at the performance of the marker combination selected
with combiroc using the combi-score, which is calculated here for the
top <em>Combination 172173</em> (made by genes GZMB, IL2RB, KLRF1, SPON2
and TRDC), which has the highest accuracy. This score expresses the
probability of being NK cells.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># adding combi score for combination 172173</span></span>
<span><span class="va">cs_cbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combi_score.html">combi_score</a></span><span class="op">(</span><span class="va">test_cbmc</span>, Models <span class="op">=</span> <span class="va">reports</span><span class="op">$</span><span class="va">Models</span>, Metrics <span class="op">=</span> <span class="va">reports</span><span class="op">$</span><span class="va">Metrics</span>, Positive_class <span class="op">=</span> <span class="st">'NK'</span>, Negative_class <span class="op">=</span> <span class="st">'Other'</span><span class="op">)</span></span>
<span><span class="va">cbmc</span><span class="op">[[</span><span class="st">'c172173_combi_score'</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span> <span class="va">cs_cbmc</span><span class="op">$</span><span class="va">`Combination 172173`</span>  </span></code></pre></div>
<p>Let’s plot the combi-score for <em>Combination 172173</em> across
CBMC clusters:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features <span class="op">=</span> <span class="st">"c172173_combi_score"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features <span class="op">=</span> <span class="st">"c172173_combi_score"</span>, group.by <span class="op">=</span> <span class="st">"ID"</span>, pt.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="cbmc_dimviolin_c172173.png" width="92160" style="display: block; margin: auto;"></p>
<p>The combi-score, which is the probability of belonging to NK cells,
is highly specific for the clusters that are annotated with a “NK”
label. pDCs cells do not show any signal and a very small number of
mouse cells are colored in the FeaturePlot (left), yet without any
relevance to the combi-score as seen in the violin plots (right).</p>
</div>
</div>
<div class="section level3">
<h3 id="testing-marker-combination-on-pbmc3k-dataset">Testing marker combination on PBMC3K dataset<a class="anchor" aria-label="anchor" href="#testing-marker-combination-on-pbmc3k-dataset"></a>
</h3>
<p>The Peripheral blood mononuclear cells (PBMC3K) test dataset can also
be directly invoked from the <code>SeuratData</code> library. The
dataset is immediately available as an already processed Seurat object
named <code>pbmc3k.final</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve pbmc3k dataset from SeuratData package</span></span>
<span><span class="co"># library(SeuratData)</span></span>
<span><span class="fu">InstallData</span><span class="op">(</span><span class="st">"pbmc3k.SeuratData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc3k.final"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k.final</span> <span class="op">&lt;-</span> <span class="va">pbmc3k.final</span></span>
<span></span>
<span><span class="co"># add cell annotations to identities</span></span>
<span><span class="va">pbmc3k.final</span><span class="op">[[</span><span class="st">'ID'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc3k.final</span><span class="op">)</span></span>
<span><span class="va">pbmc3k.final</span></span></code></pre></div>
<p>The original UMAP embedding of the PBMC3K datasets:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">T</span>, repel <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_dimplot.png" width="67200" style="display: block; margin: auto;"></p>
<p>In the same way as previously done we are using now
<code><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc()</a></code> to obtain the second combiroc test
dataset (<code>test_pbmc3k</code>). Please note that this dataset does
not contain TRDC expression values since this gene (which contributes to
the top selected combinations) is not present in the original PBMC3K
dataset.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_to_combiroc.html">seurat_to_combiroc</a></span><span class="op">(</span>SeuratObject <span class="op">=</span> <span class="va">pbmc3k.final</span>, gene_list <span class="op">=</span> <span class="va">nk_genes</span>, assay <span class="op">=</span> <span class="st">'RNA'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test_pbmc3k</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="distribution-of-individual-markers-expression-1">Distribution of individual markers’ expression<a class="anchor" aria-label="anchor" href="#distribution-of-individual-markers-expression-1"></a>
</h4>
<p>Let’s have a look how the top four single NK-markers expression is
distributed in the PBMC-3K dataset:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features <span class="op">=</span> <span class="va">nk_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_violinplots_genes.png" width="67200" style="display: block; margin: auto;"></p>
<p>In this PBMC dataset, all top four markers have high expression
values in NK cells and in CD8 T cells too, showing that individual
markers mRNA level may be not sufficient to clearly identify the cells
of interest.</p>
</div>
<div class="section level4">
<h4 id="combiroc-nk-probability-across-pbmc3k-cell-clusters">Combiroc NK probability across PBMC3K cell clusters<a class="anchor" aria-label="anchor" href="#combiroc-nk-probability-across-pbmc3k-cell-clusters"></a>
</h4>
<p>We now look at the performance of the marker combination selected
with combiroc using the combi-score, which is calculated here for the
top <em>Combination 137550</em> (made by genes CLIC3, FCGR3A, IL2RB,
KLRD1 and KLRF1), the highest accuracy combination without TRDC gene
(which is not expressed in the this dataset). This score also expresses
the probability of being NK cells.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sub</span> <span class="op">&lt;-</span> <span class="va">reports</span><span class="op">$</span><span class="va">Models</span></span>
<span><span class="va">sub</span><span class="op">$</span><span class="va">`Combination 169807`</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">sub</span><span class="op">$</span><span class="va">`Combination 172173`</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">cs_pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combi_score.html">combi_score</a></span><span class="op">(</span><span class="va">test_pbmc3k</span>, Models <span class="op">=</span> <span class="va">sub</span>,</span>
<span>    Metrics <span class="op">=</span> <span class="va">reports</span><span class="op">$</span><span class="va">Metrics</span>, Positive_class <span class="op">=</span> <span class="st">"NK"</span>, Negative_class <span class="op">=</span> <span class="st">"Other"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k.final</span><span class="op">[[</span><span class="st">"c137550_combi_score"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">cs_pbmc3k</span><span class="op">$</span><span class="va">`Combination 137550`</span>  <span class="co"># to add combi score of combination 137550</span></span></code></pre></div>
<p>Let’s plot the combi-score for <em>Combination 137550</em> across
PBMC-3K clusters:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features <span class="op">=</span> <span class="st">"c137550_combi_score"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features <span class="op">=</span> <span class="st">"c137550_combi_score"</span>, group.by <span class="op">=</span> <span class="st">"ID"</span>, pt.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="pbmc3k_dimviolin_c137550.png" width="92160" style="display: block; margin: auto;"></p>
<p>Here too, as seen for the previous dataset, the cell cluster labeled
as “NK” is the only one displaying a very high combi-score, i.e. showing
an actual probability of being NK cells. No other cell in any cluster
produces noisy interference.</p>
</div>
</div>
<div class="section level3">
<h3 id="combinations-selected-with-combiroc-are-optimised-smaller-signatures">Combinations selected with combiroc are optimised smaller
signatures<a class="anchor" aria-label="anchor" href="#combinations-selected-with-combiroc-are-optimised-smaller-signatures"></a>
</h3>
<p>While the <strong>combi-score</strong> described here is defined as
the predicted probability obtained while fitting the combinations models
to the test datasets (returned by <code><a href="../reference/combi_score.html">combi_score()</a></code>), other
metrics exist such as the <strong>gene-signature-score</strong>
described in Della Chiara et al. 2021 <a href="https://www.nature.com/articles/s41467-021-22544-y" class="external-link">[5]</a>: the
gene-signature-score takes into account both the expression level and
co-expression of genes within each single cell.<br>
Given a geneset, the gene-signature-score’s increase is directly
proportional to the number of expressed genes in the signature and to
the sum of their level of expression.</p>
<p>We are going now to use this gene-signature-score computed on the
whole signature and computed on the combiroc-selected combinations. To
do so, we reproduced the gene-signature-score computation in the custom
R function <code>signature_score.R</code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"inst/external_code/signature_score.R"</span><span class="op">)</span></span></code></pre></div>
<p>First, we compute the “gene-signature-score” for the <strong>whole 30
DEGs signature</strong> that we assume as reference. Being this a whole
signature (30 genes) we expect it to be very specific. We are doing this
on both test datasets previously used.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># computing whole signature gene-signature-score for CBMC</span></span>
<span><span class="va">NK_sig_cbmc</span><span class="op">&lt;-</span><span class="fu">signature_score</span><span class="op">(</span>SeuratObj <span class="op">=</span> <span class="va">cbmc</span>, geneset <span class="op">=</span> <span class="va">nk_genes</span><span class="op">)</span></span>
<span><span class="va">cbmc</span><span class="op">$</span><span class="va">NK_signature_score</span> <span class="op">&lt;-</span> <span class="va">NK_sig_cbmc</span><span class="op">$</span><span class="va">scaled_combined_score</span></span>
<span></span>
<span><span class="co"># computing whole signature gene-signature-score for PBMC-3K</span></span>
<span><span class="va">NK_sig_pbmc3k</span><span class="op">&lt;-</span><span class="fu">signature_score</span><span class="op">(</span>SeuratObj <span class="op">=</span> <span class="va">pbmc3k.final</span>, geneset <span class="op">=</span> <span class="va">nk_genes</span><span class="op">)</span></span>
<span><span class="va">pbmc3k.final</span><span class="op">$</span><span class="va">NK_signature_score</span> <span class="op">&lt;-</span> <span class="va">NK_sig_pbmc3k</span><span class="op">$</span><span class="va">scaled_combined_score</span></span></code></pre></div>
<p>Then, we compute the “gene-signature-score” <strong>limited to the
combinations of five genes previously used</strong>: - <em>Combination
172173</em> (made by genes GZMB, IL2RB, KLRF1, SPON2 and TRDC) in CBMC
dataset. - <em>Combination 137550</em> (made by genes CLIC3, FCGR3A,
IL2RB, KLRD1 and KLRF1) in PBMC3K dataset These combinations are going
to be our “optimized-signatures”.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># computing gene-signature-score of combination 172173 for CBMC</span></span>
<span><span class="va">comb_cbmc</span><span class="op">&lt;-</span><span class="fu">signature_score</span><span class="op">(</span>SeuratObj <span class="op">=</span> <span class="va">cbmc</span>, geneset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'GZMB'</span>,<span class="st">'IL2RB'</span>,<span class="st">'KLRF1'</span>,<span class="st">'SPON2'</span>,<span class="st">'TRDC'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cbmc</span><span class="op">[[</span><span class="st">'c172173_signature_score'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">comb_cbmc</span><span class="op">$</span><span class="va">scaled_combined_score</span>  </span>
<span></span>
<span><span class="co"># computing gene-signature-score of combination 137550 for PBMC-3K</span></span>
<span><span class="va">comb_pbmc3k</span><span class="op">&lt;-</span><span class="fu">signature_score</span><span class="op">(</span>SeuratObj <span class="op">=</span> <span class="va">pbmc3k.final</span>, geneset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'CLIC3'</span>, <span class="st">'FCGR3A'</span>, <span class="st">'IL2RB'</span>,<span class="st">'KLRD1'</span>,<span class="st">'KLRF1'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc3k.final</span><span class="op">[[</span><span class="st">'c137550_signature_score'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">comb_pbmc3k</span><span class="op">$</span><span class="va">scaled_combined_score</span> </span></code></pre></div>
<p>We are now plotting side by side the gene-signature-scores from the
whole-signature (30 genes) and the combinations (5 genes).<br>
For the CBMC dataset:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features<span class="op">=</span><span class="st">'NK_signature_score'</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features<span class="op">=</span><span class="st">'c172173_signature_score'</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v1</span><span class="op">|</span><span class="va">v2</span></span></code></pre></div>
<p><img src="cbmc_NK_c172173_sigscore.png" width="92160" style="display: block; margin: auto;"></p>
<p>And for the PBMC-3K dataset:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features<span class="op">=</span><span class="st">'NK_signature_score'</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features <span class="op">=</span> <span class="st">"c137550_signature_score"</span>, group.by <span class="op">=</span> <span class="st">"ID"</span>, pt.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v1</span><span class="op">|</span><span class="va">v2</span></span></code></pre></div>
<p><img src="pbmc3k_NK_c137550_sigscore.png" width="92160" style="display: block; margin: auto;"></p>
<p>As expected, the gene-signature-score for both datasets correctly
discriminates NK-cells from others, but while in CBMC cells the NK
cluster is unequivocally selected, in PBMC-3K cells also cytotoxic CD8-T
cells are picked with a still relevant score. This does not happen when
the gene-signature-score is computed over the combinations selected with
combiroc (plots on the right). Overall, the gene-signature-scores on
combinations are highly specific and less noisy.</p>
<p>This demonstrates that having more genes as in the 30-genes whole
signature does not add discriminatory power compared to combinations
with much less genes. The 5-genes <strong>combinations selected with
combiroc are de-facto optimized gene signatures</strong> with equal or
even higher discriminatory power compared to the bigger gene signatures
they are extracted from.</p>
</div>
<div class="section level3">
<h3 id="the-combi-score-can-be-used-to-discriminate-among-cell-clusters">The combi-score can be used to discriminate among cell clusters<a class="anchor" aria-label="anchor" href="#the-combi-score-can-be-used-to-discriminate-among-cell-clusters"></a>
</h3>
<p>Finally we compare here the complete NK whole gene-signature-score
with the combiroc’s NK combi-score to see how they perform in both test
datasets.</p>
<p>Comparison in CBMC cells:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features<span class="op">=</span><span class="st">'NK_signature_score'</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features<span class="op">=</span><span class="st">'c172173_combi_score'</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v1</span><span class="op">|</span><span class="va">v2</span></span></code></pre></div>
<p><img src="cbmc_NK_c172173_combiscore.png" width="92160" style="display: block; margin: auto;"></p>
<p>Comparison in PBMC-3K cells:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features<span class="op">=</span><span class="st">'NK_signature_score'</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features <span class="op">=</span> <span class="st">"c137550_combi_score"</span>, group.by <span class="op">=</span> <span class="st">"ID"</span>, pt.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">v1</span><span class="op">|</span><span class="va">v2</span></span></code></pre></div>
<p><img src="pbmc3k_NK_c137550_combiscore.png" width="92160" style="display: block; margin: auto;"></p>
<p>Remarkably, despite the 6 fold reduction in the number of considered
features to build the gene score, for both the datasets, the signature
score confirmed that the best combination is, at worst, as accurate as
the whole signature in indicating the NK cluster.<br>
In addition, combi-score results to be even more precise and further
suggests that these combinations can be used as a refined version of the
30-genes NK whole-signature.</p>
<p>The difference between the gene-signature-score and combi-score is
not surprising since to compute the first the genes within the
combinations are considered equally, while in the latter combiroc models
assign a ‘weight’ to each gene, further indicating the power of combiroc
combination models in suggesting the identity of a cluster with a modest
amount of genes.</p>
<p>As a mean of control, we check the effect of a set of random
combinations on the test datasets</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combs_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">random_combs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1492</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">combs_length</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">random_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cbmc</span><span class="op">)</span>, <span class="va">i</span><span class="op">)</span></span>
<span>      <span class="va">dfc</span> <span class="op">&lt;-</span> <span class="fu">signature_score</span><span class="op">(</span><span class="va">cbmc</span>, <span class="va">random_list</span><span class="op">)</span></span>
<span>      <span class="va">cbmc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"random_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">dfc</span><span class="op">$</span><span class="va">scaled_combined_score</span></span>
<span>      <span class="va">dfp</span> <span class="op">&lt;-</span> <span class="fu">signature_score</span><span class="op">(</span><span class="va">pbmc3k.final</span>, <span class="va">random_list</span><span class="op">)</span></span>
<span>      <span class="va">pbmc3k.final</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"random_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">dfp</span><span class="op">$</span><span class="va">scaled_combined_score</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Here random combinations do not have a preferential cluster and show
a generalized higher background signal noise. For CBMC and PBMC-3K
datasets:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">cbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'random_30'</span>, <span class="st">'random_5'</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="cbmc_violin_randomcontrol.png" width="73728" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k.final</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'random_30'</span>, <span class="st">'random_5'</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">'ID'</span>, pt.size<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc3k_violin_randomcontrol.png" width="73728" style="display: block; margin: auto;"></p>
<p>Beyond using scores of random combinations as negative control, they
also show the magnitude of background noise associated to each random
combination taken into consideration. In fact, random scores show quite
homogeneously distributed values across clusters indicating that the
cluster-preferential distributions of real combinations scores are not
stochastic.<br>
Generally speaking since the higher is the length of random signatures,
the higher will be the noise hence the probability to pick genes
expressed in all the clusters: this suggests that <strong>the higher is
the length of a real DEG signature the higher is the probability to
include less specific genes and to increase the noise</strong>.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>This guided tutorial shows how to use the combiroc package to
optimize single-cell gene signature with combinations of marker and to
identify a population of interest in scRNA-seq datasets:</p>
<ul>
<li>combiroc accurately selects the most informative markers from a
wider trancriptional signature.</li>
<li>combiroc’s specificity can also ameliorate results obtained by other
published gene-signature scores: the gene-signature summarizing method
is remarkably less noisy when performed on markers combinations compared
to whole gene signatures.</li>
</ul>
<p>Identifying NK cells using the expression level of the best markers
individually is not efficient since they are too unspecific for
classification purposes if taken one by one.<br>
The power of a combined signature of five markers only is comparable -
and in some cases superior - to that of a whole scRNA-seq gene
signature, typically in the range of tens of genes: <strong>scaling down
a gene signature while retaining its discriminatory power is highly
relevant in both research and diagnostic context</strong>.</p>
<p>Back to the <a href="#top">top</a> of this docment.</p>
<p>Go to <a href="https://ingmbioinfo.github.io/combiroc/articles/combiroc_vignette_1.html">Combiroc
standard workflow tutorial</a></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Abdelaal et al. 2019. <em>Genome Biology</em>. A comparison of
automatic cell identification methods for single-cell RNA sequencing
data</li>
<li>Hao et al. 2021.<em>Cell</em> Integrated analysis of multimodal
single-cell data</li>
<li>Stoeckius et al. 2017. <em>Nature Methods</em>. Simultaneous epitope
and transcriptome measurement in single cells</li>
<li>Satija et al. 2015. <em>Nature Biotechnology</em>. Spatial
reconstruction of single-cell gene expression data.</li>
<li>Della Chiara et al. 2021. <em>Nature Communications</em>. Epigenomic
landscape of human colorectal cancer unveils an aberrant core of
pan-cancer enhancers orchestrated by YAP/TAZ.</li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ivan Ferrari, Riccardo L. Rossi, Saveria Mazzara.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
